client.go

```diff
@@ -98,10 +98,11 @@ func readExpr() <-chan expr {
 			log.Printf("recv: %s", s.Text())
 			if word, rest := splitWord(s.Text()); word == "recv" {
 				gState.mutex.Lock()
```

Comment:
(Non-blocking, we can fix it and/or continue to discuss this after the PR is merged) Should we maybe error out if we can't lock the mutex in some amount of time?

The obvious function we'd use to implement this comes with a warning, though (https://pkg.go.dev/sync#Mutex.TryLock). Perhaps we are meant to use channels instead.

Reply:
I think the reason I didn't use a channel was because it's asynchronous, so theoretically it's possible for the `readExpr` thread to receive the query and respond with the old state before obtaining the update, though I think it would be unlikely.

Anyway I don't see this being a problem because the main thread is supposed to hold the mutex for only a very short amount of time when it is exporting the state in the `runShell` function.